### PR TITLE
migrate some material files to nullsafety

### PR DESCRIPTION
**Pull request by a14n**
## Description

NNBD migration for part of material

## Related Issues

* dart-lang/sdk#40146

## Tests

I added the following tests:

None

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/animated_icons/animated_icons.dart [hunk 1]
```diff
@@ -92,26 +90,26 @@ class AnimatedIcon extends StatelessWidget {
...
   @override
   Widget build(BuildContext context) {
     final _AnimatedIconData iconData = icon as _AnimatedIconData;
     final IconThemeData iconTheme = IconTheme.of(context);
-    final double iconSize = size ?? iconTheme.size;
-    final TextDirection textDirection = this.textDirection ?? Directionality.of(context);
-    final double iconOpacity = iconTheme.opacity;
-    Color iconColor = color ?? iconTheme.color;
+    final double iconSize = size ?? iconTheme.size!;
+    final TextDirection textDirection = this.textDirection ?? Directionality.of(context)!;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/animated_icons/animated_icons.dart:110]: Add a debugCheckHasDirectionality.

File: packages/flutter/lib/src/material/animated_icons/animated_icons.dart [hunk 2]
```diff
@@ -92,26 +90,26 @@ class AnimatedIcon extends StatelessWidget {
...
 
   @override
   Widget build(BuildContext context) {
     final _AnimatedIconData iconData = icon as _AnimatedIconData;
     final IconThemeData iconTheme = IconTheme.of(context);
-    final double iconSize = size ?? iconTheme.size;
-    final TextDirection textDirection = this.textDirection ?? Directionality.of(context);
-    final double iconOpacity = iconTheme.opacity;
-    Color iconColor = color ?? iconTheme.color;
+    final double iconSize = size ?? iconTheme.size!;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/animated_icons/animated_icons.dart:109]: These `!` on the iconTheme property are unfortunate. We know that `IconTheme.of` only returns theme data where that is true, I wish we could express that in code better. 

Maybe add an `assert(iconTheme.isConcrete)` above this line to make it clear to the reader?

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart [hunk 3]
```diff
@@ -140,27 +138,27 @@ class CalendarDatePicker extends StatefulWidget {
...
-  DateTime _selectedDate;
+  DatePickerMode? _mode;
+  DateTime? _currentDisplayedMonthDate;
+  late DateTime _selectedDate;
   final GlobalKey _monthPickerKey = GlobalKey();
   final GlobalKey _yearPickerKey = GlobalKey();
-  MaterialLocalizations _localizations;
-  TextDirection _textDirection;
+  late MaterialLocalizations _localizations;
+  TextDirection? _textDirection;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/pickers/calendar_date_picker.dart:161]: This could be late, I think?

**Review by goderbauer — COMMENTED** [review]
I reviewed everything outside the `picker` directory. All that LGTM.

@darrenaustin I think you're more familiar with the datepicker code. Could you review the files inside the `picker` directory? Thanks.

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart (hunk 3 above)
Comment (a14n) [reply packages/flutter/lib/src/material/pickers/calendar_date_picker.dart]: You're right, thanks.

File: packages/flutter/lib/src/material/animated_icons/animated_icons.dart (hunk 2 above)
Comment (a14n) [reply packages/flutter/lib/src/material/animated_icons/animated_icons.dart:111]: assert added

File: packages/flutter/lib/src/material/animated_icons/animated_icons.dart (hunk 1 above)
Comment (a14n) [reply packages/flutter/lib/src/material/animated_icons/animated_icons.dart:112]: added

File: packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart [hunk 4]
```diff
@@ -38,14 +35,14 @@ const double _maxCalendarWidthPortrait = 480.0;
...
+    DateTime? initialStartDate,
+    DateTime? initialEndDate,
+    required DateTime firstDate,
+    required DateTime lastDate,
+    DateTime? currentDate,
+    required this.onStartDateChanged,
+    required this.onEndDateChanged,
   }) : initialStartDate = initialStartDate != null ? utils.dateOnly(initialStartDate) : null,
        initialEndDate = initialEndDate != null ? utils.dateOnly(initialEndDate) : null,
        assert(firstDate != null),
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart:48]: Do we still need this assert (and the one below it)? These will never be null, right?

File: packages/flutter/lib/src/material/pickers/date_picker_dialog.dart [hunk 5]
```diff
@@ -102,26 +100,26 @@ const double _inputFormLandscapeHeight = 108.0;
...
-  String errorFormatText,
-  String errorInvalidText,
-  String fieldHintText,
-  String fieldLabelText,
+  String? errorFormatText,
+  String? errorInvalidText,
+  String? fieldHintText,
+  String? fieldLabelText,
 }) async {
   assert(context != null);
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:124]: These asserts are no longer needed.

File: packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart [hunk 6]
```diff
@@ -109,28 +107,28 @@ const double _inputFormLandscapeHeight = 108.0;
...
+  Locale? locale,
   bool useRootNavigator = true,
-  RouteSettings routeSettings,
-  TextDirection textDirection,
-  TransitionBuilder builder,
+  RouteSettings? routeSettings,
+  TextDirection? textDirection,
+  TransitionBuilder? builder,
 }) async {
     assert(context != null);
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/pickers/date_range_picker_dialog.dart:133]: Assert no longer required.

**Review by darrenaustin — APPROVED** [review]
Just reviewing the picker directory, this LGTM.

Just a few asserts that are no longer needed, but if you can't get to them with this checkin, I can remove them later. Thanks for all the work here!

**Comment by goderbauer** [comment]
For the time being (since most people will run this code without null-safety enabled) we still need the asserts.

File: packages/flutter/lib/src/material/pickers/calendar_date_picker.dart [hunk 7]
```diff
@@ -87,12 +85,12 @@ class CalendarDatePicker extends StatefulWidget {
...
   /// [initialDate].
   CalendarDatePicker({
-    Key key,
-    @required DateTime initialDate,
-    @required DateTime firstDate,
-    @required DateTime lastDate,
-    DateTime currentDate,
-    @required this.onDateChanged,
+    Key? key,
+    required DateTime initialDate,
```
[the remark this answers is not in the document]
Comment (a14n) [reply packages/flutter/lib/src/material/pickers/calendar_date_picker.dart:89]: > This is something we can relax after this work is complete without breaking existing apps, right?

In the case of parameters it is not breaking to change from required non-nullable to accept nullable.

**Comment by a14n** [comment]
Regarding your assert comment, a non migrated code can still pass null to a non nullable parameter. So assert are still useful.

**Pull request merged**